{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 04 Assignment genes expressed in the brain\n",
    "\n",
    "## Introduction\n",
    "\n",
    "During this week, you will learn how to define your own classes and use them to create objects. In Python everything is an object and even without knowing it we have been using these\n",
    "objects. \n",
    "\n",
    "### objects and classes \n",
    "\n",
    "We can recognize an object by the **dot** (**.**) notation that we have to use to get access to the functions and variables of an object. In object terminology a class function is called a method and a class variable is called an attribute.\n",
    "\n",
    "For example, remember the string.upper() method? This is an example of calling the upper() method on the string object. We can see that we used the dot operator after the name of the object we want to use and that the method will work in the variables that are saved in the object (called the object state).\n",
    "\n",
    "In general we can say that objects have a **state** (variables or **attributes** called in objects) and **behavior** (functions or **methods** called in objects) that work on these variables).\n",
    "\n",
    "Objects are defined in so called **classes**, which are the blueprint of the object to be created. Object can be called into live (object instantiation) by assigning this blueprint to a variable.\n",
    "\n",
    "For the assignment of this week you wil define two class definitions and use these to bring objects of these types into life by assigning it to a variable.\n",
    "\n",
    "### Research context: Studying Gene Expression in Different Tissue (regions). \n",
    "\n",
    "Imagine you're a researcher studying the gene expression patterns in different tissues of a model organism. You have conducted a microarray experiment where you've collected gene expression data from various tissues or tissue regions, and you want to analyze the data to identify genes that are highly expressed in specific tissues (regions).\n",
    "\n",
    "For instance, you have collected data from two tissues regions in the brain: \"LHM,\" and \"PHA\".  Each tissue location has its own set of gene expression values measured by probes. The probes provide information about the expression level of individual genes.\n",
    "\n",
    "By using python MicroArray class and a Probe class you provide structured code to analyze the results of such an experiment. \n",
    "\n",
    "You can use the MicroArray class to create microarray objects for each tissue region (in our example two microarray objects, one for LHM and one for PHA), and each object would store the tissue-region-specific gene expression data. You can create instances of the Probe class to represent the gene expression values for each gene in each tissue-region.\n",
    "\n",
    "With the MicroArray class, you can:\n",
    "-\tCreate microarray objects for each tissue, storing the tissue's ID, acronym, and name, along with the gene expression data.\n",
    "-\tUse a method to filter out probes with expression levels above a certain cutoff value. This can help you identify genes that are highly expressed in a specific tissue. \n",
    "-\tOptionally filter for background signals.\n",
    "-\tKeep track of the total number of microarray experiments conducted.\n",
    "\n",
    "After filtering the probes based on the expression cutoff, and optional filtering the background noise, you can analyze the resulting list of probes to identify genes that are likely to be important in the context of each tissue. This information could lead to insights into tissue-specific gene functions and the biological mechanism. \n",
    "\n",
    "**about the background noise**\n",
    "When molecules bind to probes on a microarray chip, they generate a signal that is detected and measured. However, not all signal detected is necessarily due to the specific gene expression being measured. There could be various sources of noise or non-specific binding that contribute to the overall signal, and this is what is referred to as the \"background\" signal. As a researcher, you might want to select only the expression values that are above a background signal level. \n",
    "Source: Jaksik R, Iwanaszko M, Rzeszowska-Wolny J, Kimmel M. Microarray experiments and factors which affect their reliability. Biol Direct. 2015 Sep 3;10:46. doi: 10.1186/s13062-015-0077-2. PMID: 26335588; PMCID: PMC4559324.\n",
    "\n",
    "In summary, with the MicroArray and Probe classes you create structured code to manage and analyze gene expression data from microarray experiments, helping researchers make sense of complex molecular biology datasets.\n",
    "\n",
    "Files needed: `MicroarrayExpression.csv`, `Probes.csv`, `SampleAnnot.csv` and `PACall.csv`\n",
    "\n",
    "## Class creation instructions:\n",
    "\n",
    "Create a program that will read all files needed and creates objects of the following types:\n",
    "\n",
    "### MicroArray class\n",
    " - should store the Structure id, Structure acronym, Structure name\n",
    " - has a list of probe objects for a given experiment\n",
    " - above information is mandatory to pass when creating a microarray object\n",
    " - should keep track of the number of created microarrays\n",
    " - when printing a microarray object it should show the Structure id, Structure acronym and Structure name nicelly formatted\n",
    " - has a method to get the probes that are above a certain expression value (cutoff) \n",
    "\n",
    "### Probe class\n",
    "- should store expression value\n",
    "- should store probe_id, gene id and gene name, chromosome\n",
    "- should store if the probe is above background (see `PACall.csv` file)\n",
    "- it should be mandatory to pass all information when the probe is created\n",
    "- when a probe is printed it should show a nicelly formatted string containing the value of the attributes\n",
    "- should keep track of the number of created probes\n",
    "\n",
    "Run your program from the commandline and show the difference and intersection of two experiments (i.e. LHM and PHA), even beter would be to get two acronyms from the commandline and use these.\n",
    "\n",
    "## Learning outcomes:\n",
    "- know how to define and use classes; \n",
    "    - including the hooks `__init__` and  `__str__` \n",
    "    - know the difference between an instance attribute and class attribute\n",
    "- Can use `try except` statements correctly\n",
    "- write your own error classes\n",
    "\n",
    "\n",
    "## Coding requirements:\n",
    "- write a Python script ending on `.py` (not a notebook), that can be called from the commandline while giving it all necessary information (files and cutoff)\n",
    "- when designing classes; think of which attributes and methods belong to a certain class\n",
    "- when designing functions; keep them simple and have them (preferentially) do only one thing. What a function does should also be clear from the name. \n",
    "- Every module, class and function should have pydoc (see PEP8)\n",
    "- Your program should follow the PEP8 standard. On our system you can use the commandline tool `pylint` to check this. Run it in the terminal by typing: `pylint name_of_your_program.py`. Visual Code has extensions (pylint, Flake8) that allow you to see the PEP8 warnings inside the IDE.\n",
    "- add `try except` statements where appropriate\n",
    "    - define your own error class and `raise` it \n",
    "- Place the class definitions in a seperate module and import it.\n",
    "- define at least two additional classes (see above): `MicroArray` and `Probe`\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "vscode": {
   "interpreter": {
    "hash": "b0fa6594d8f4cbf19f97940f81e996739fb7646882a419484c72d19e05852a7e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
